### PR TITLE
Traffic Manager 2017-03-01 spec

### PR DESCRIPTION
**Pull request by vladca**
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x ] The title of the PR is clear and informative.
- [ x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ x] If applicable, the PR references the bug/issue that it fixes.
- [ x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x ] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [ x] My spec meets the review criteria:
  - [ x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x ] Validation errors from the [Linter extension for VS Code](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/linter.md) have all been fixed for this spec. (**Note:** for large, previously checked in specs, there will likely be many errors shown. Please contact our team so we can set a timeframe for fixing these errors if your PR is not going to address them).
  - [ x] The spec follows the patterns described in the [Swagger good patterns](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-good-patterns.md) document unless the service API makes this impossible.

### CODE REVIEW

**Comment by vladca** [comment]
It's not really helpful for diffing that we add a new file of each API version. Perhaps I should have committed the 2015-11-01 file in the 2017-03-01 folder and then change it to make it easier. Sorry about that. I will highlight the changes for reviewers

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json [hunk 1]
```diff
@@ -0,0 +1,845 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "title": "TrafficManagerManagementClient",
+    "version": "2017-03-01"
```
Comment (vladca) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: Changed

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json [hunk 2]
```diff
@@ -0,0 +1,845 @@
...
+          "200": { 
+             "description": "The updated Traffic Manager profile.", 
+             "schema": { 
+               "$ref": "#/definitions/Profile" 
+             } 
+          }
+        }
+      }
+    },
+    "/providers/Microsoft.Network/trafficManagerGeographicHierarchies/default": {
```
Comment (vladca) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: Added new

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json [hunk 3]
```diff
@@ -0,0 +1,845 @@
...
+        "endpointMonitorStatus": {
+          "type": "string",
+          "description": "Gets or sets the monitoring status of the endpoint."
+        },
+        "minChildEndpoints": {
+          "type": "integer",
+          "format": "int64",
+          "description": "Gets or sets the minimum number of endpoints that must be available in the child profile in order for the parent profile to be considered available. Only applicable to endpoint of type 'NestedEndpoints'."
+        },
+        "geoMapping": {
```
Comment (vladca) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: Added new

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json [hunk 4]
```diff
@@ -0,0 +1,845 @@
...
+          "description": "The reason why the name is not available, when applicable."
+        },
+        "message": {
+          "type": "string",
+          "description": "Descriptive message that explains why the name is not available, when applicable."
+        }
+      },
+      "description": "Class representing a Traffic Manager Name Availability response."
+    },
+    "Region": {
```
Comment (vladca) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: Added new

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json [hunk 5]
```diff
@@ -0,0 +1,845 @@
...
+    "GeographicHierarchyProperties": {
+      "properties": {
+        "geographicHierarchy": {
+          "$ref": "#/definitions/Region",
+          "description": "The region at the root of the hierarchy from all the regions in the hierarchy can be retrieved."
+        }
+      },
+      "description": "Class representing the properties of the Geographic hierarchy used with the Geographic traffic routing method."
+    },
+    "TrafficManagerGeographicHierarchy": {
```
Comment (vladca) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: Added new

**Comment by sarangan12** [comment]
I am starting my review of this swagger. I will update the details shortly

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json (hunk 5 above)
Comment (sarangan12) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: This is a TrackedResource i.e. extends Resource (using allOf). There must be a Patch operation for this tracked resource that atleast updates the tags. Could you add it?

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json [hunk 6]
```diff
@@ -0,0 +1,845 @@
...
+      "properties": {
+        "name": {
+          "type": "string",
+          "description": "The relative name."
+        },
+        "type": {
+          "type": "string",
+          "description": "Traffic Manager profile resource type."
+        },
+        "nameAvailable": {
```
Comment (sarangan12) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: It is not recommended to use type: boolean. Instead, the recommended approach is to use string enums with allowed set of values defined. Could you modify it?

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json [hunk 7]
```diff
@@ -0,0 +1,845 @@
...
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/trafficmanagerprofiles": {
+      "get": {
+        "tags": [
+          "Profiles"
+        ],
+        "x-ms-pageable": {
+          "nextLinkName": null
+        },
+        "operationId": "Profiles_ListAllInResourceGroup",
```
Comment (sarangan12) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: The recommended naming convention is "Profiles_ListByResourceGroup". Can this be changed?

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json [hunk 8]
```diff
@@ -0,0 +1,845 @@
...
+               "$ref": "#/definitions/Endpoint" 
+             } 
+          } 
+        }
+      },
+      "get": {
+        "tags": [
+          "Endpoints"
+        ],
+        "operationId": "Endpoints_Get",
```
Comment (sarangan12) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: Does this name sound right? Endpoint**s**_Get or Endpoint_Get?

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json [hunk 9]
```diff
@@ -0,0 +1,845 @@
...
+               "$ref": "#/definitions/Endpoint" 
+             } 
+          } 
+        }
+      },
+      "put": {
+        "tags": [
+          "Endpoints"
+        ],
+        "operationId": "Endpoints_CreateOrUpdate",
```
Comment (sarangan12) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: Same here. All the operations deal with a single endpoint. But the names are misleading as Endpoints. Can I actually create or update multiple endpoints?

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json [hunk 10]
```diff
@@ -0,0 +1,845 @@
...
+            "description": "The Traffic Manager endpoint parameters supplied to the Update operation."
+          },
+          {
+            "$ref": "#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "#/parameters/SubscriptionIdParameter"
+          }
+        ],
+        "responses": {
```
Comment (sarangan12) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: For all the responses, there is no default response. I agree that in the absence of a default response, autorest would use the standard cloud exception. But, we are actually recommending RPs to provide a default error response if possible. Please try to add a default response for all the responses.

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json [hunk 11]
```diff
@@ -0,0 +1,845 @@
...
+          "200": { 
+             "description": ""
```
Comment (sarangan12) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: Empty description here and in the next response. Please fix it.

Comment (sarangan12) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: WHat is the schema that is returned?

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json [hunk 12]
```diff
@@ -0,0 +1,845 @@
...
+          {
+            "$ref": "#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "#/parameters/SubscriptionIdParameter"
+          }
+        ],
+        "responses": {
+          "200": { 
+             "description": "",
```
Comment (sarangan12) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: Empty description

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json [hunk 13]
```diff
@@ -0,0 +1,845 @@
...
+          {
+            "$ref": "#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "#/parameters/SubscriptionIdParameter"
+          }
+        ],
+        "responses": {
+          "200": { 
+             "description": ""
```
Comment (sarangan12) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: Empty description and no schema????

**Review by sarangan12 — CHANGES_REQUESTED** [review]

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json (hunk 1 above)
Comment (ms-alguerra) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: I believe this is a namespace broad operation and currently this operation is implemented by NRP team (although I am not sure if they expose it in Swagger)

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json (hunk 5 above)
Comment (ms-alguerra) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: I believe this operation is exposed as proxy-only in the manifest. Will removing the allOf bit be enough to fix it?

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json (hunk 8 above)
Comment (ms-alguerra) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: I believe the reason for this is consistency with the URI naming (plural is used there across RPs):

/subscriptions/{subscriptionId}/providers/Microsoft.Network/trafficmanagerprofiles

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json (hunk 9 above)
Comment (ms-alguerra) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: Same comment about consistency with URI definition

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json (hunk 11 above)
Comment (ms-alguerra) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: Will do

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json (hunk 12 above)
Comment (ms-alguerra) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: Will fix

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json (hunk 7 above)
Comment (ms-alguerra) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: Will fix

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json (hunk 6 above)
Comment (ms-alguerra) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: That will require server side change so it is not as straight forward. For now it is a won't fix.

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json [hunk 14]
```diff
@@ -0,0 +1,845 @@
...
+        }
+      },
+      "allOf": [
+        {
+          "$ref": "#/definitions/Resource"
+        }
+      ],
+      "description": "Class representing the Geographic hierarchy used with the Geographic traffic routing method."
+    },
+    "Resource": {
```
Comment (ms-alguerra) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: Will do

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json (hunk 10 above)
Comment (ms-alguerra) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: Will do using dns spec as a template.

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json (hunk 13 above)
Comment (ms-alguerra) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: I haven't seen many teams adding schema for delete operation results. I will add one anyways.

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json (hunk 1 above)
Comment (sarangan12) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: No changes required here.

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json (hunk 5 above)
Comment (sarangan12) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: Since this is a proxy resource, no change required here,

Comment (vladca) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: This is a proxy resource and read-only, PUT and PATCH are not supported on it at all

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json [hunk 15]
```diff
@@ -0,0 +1,845 @@
...
+    "https"
+  ],
+  "consumes": [
+    "application/json",
+    "text/json"
+  ],
+  "produces": [
+    "application/json",
+    "text/json"
+  ],
```
Comment (vladca) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: I've verified that we don't support text/json and will remove the entry

**Review by sarangan12 — APPROVED** [review]

**Pull request merged**